lnrpc/routerrpc/forward_interceptor.go

```diff
@@ -0,0 +1,259 @@
+
+	var wg sync.WaitGroup
+	quit := make(chan struct{})
+	defer func() {
+		close(quit)
+		s.onInterceptorDisconnected(rpcInterceptor)
+		wg.Wait()
+	}()
+
+	// this go routine reads from the client stream and and then signal to the
+	// main loop to handle the packet.
+	clientResolutionsChan := make(chan *ForwardHtlcInterceptResponse)
+	errChan := make(chan error)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				errChan <- err
+				return
+			}
+
+			// Now that we have the response from the RPC client, send it to
+			// the responses chan.
+			select {
+			case clientResolutionsChan <- resp:
+			case <-quit:
+				return
+			case <-s.quit:
+				return
+			}
+		}
+	}()
+
+	// Our interceptor makes sure we hold the packet and then signal to the
+	// main loop to handle the packet. We only return true if we were able
+	// to deliver the packet to the main loop.
+	interceptedPacketsChan := make(chan htlcswitch.InterceptedForward)
+	fromSwitchInterceptor := func(packet htlcswitch.InterceptedForward) bool {
+		log.Infof("holding intercepted packet %v", packet)
+		rpcInterceptor.hold(packet)
+		select {
+		case interceptedPacketsChan <- packet:
+			return true
+		case <-quit:
+			return false
+		case <-s.quit:
+			return false
+		}
+	}
+
+	// Register our interceptor and make sure we resolve the held packets
+	// on disconnect.
+	interceptableForwarder := s.cfg.RouterBackend.InterceptableForwarder
+	interceptableForwarder.SetInterceptor(fromSwitchInterceptor)
+
+	for {
+		select {
+		case intercepted := <-interceptedPacketsChan:
+			log.Tracef("sending intercepted packet to client %v", intercepted)
+			// in case we couldn't forward we exit the loop and drain the
+			// current interceptor as this indicates on a connection problem.
+			if err := rpcInterceptor.forwardToClient(intercepted); err != nil {
+				return err
+			}
+		case resolution := <-clientResolutionsChan:
+			log.Tracef("resolving intercepted packet %v", resolution)
+			// in case we couldn't resolve we just add a log line since this
+			// does not indicate on any connection problem.
+			if err := rpcInterceptor.resolveFromClient(resolution); err != nil {
+				log.Warnf("client resolution of intercepted "+
+					"packet failed %v", err)
+			}
+		case err := <-errChan:
+			log.Errorf("Received an error: %v, shutting down", err)
+			return err
+		case <-s.quit:
+			return nil
+		}
+	}
+}
+
+func (s *Server) setRPCInterceptor(interceptor *rpcInterceptor) error {
+	s.rpcInterceptorMtx.Lock()
+	defer s.rpcInterceptorMtx.Unlock()
+	if s.currentRPCInterceptor != nil {
+		return ErrInterceptorAlreadyExists
+	}
+	rpcInterceptor := interceptor
+	s.currentRPCInterceptor = rpcInterceptor
+	return nil
+}
+
+// onInterceptorDisconnected removes all previousely held forwards from
+// the store. Before they are removed it ensure to resume as the default
+// behavior.
+func (s *Server) onInterceptorDisconnected(interceptor *rpcInterceptor) {
```

Comment:
interceptor is available as a field on `s`

Reply:
Yes, used that.